source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class DirectStats;
+template <typename StatsStructType> class LazyInit;
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class LazyCompatibleInterface {
+public:
+  // Helper function to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* instantiate() = 0;
+
+  virtual ~LazyCompatibleInterface() = default;
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ * This is used when
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>` is enabled.
+ */
+template <typename StatsStructType>
+class LazyInit : public LazyCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
+      : initialized_([&scope]() -> Gauge& {
+          Stats::StatNamePool pool(scope->symbolTable());
+          return Stats::Utility::gaugeFromElements(
+              *scope, {pool.add(StatsStructType::typeName()), pool.add("initialized")},
+              Stats::Gauge::ImportMode::Accumulate);
+        }()),
+        ctor_([scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          initialized_.inc();
+          return new StatsStructType(stat_names, *scope);
+        }) {
+    if (initialized_.value() > 0) {
+      instantiate();
+      ctor_ = nullptr;
+    }
+  }
+  ~LazyInit() {
+    if (initialized_.value() > 0) {
```

Comment:
per discussion this is a bug. You should null ctor_ on line 51, and use that to check whether initialized_ should be decremented.

Reply:
please refresh. 
the dec() should be coupled with ctor() invoke, updated.